intruder_alarm/src/singly/mod.rs

```diff
@@ -0,0 +1,220 @@
+//! like the allocator implementation itself, since each list element manages
+//! its own memory.
+use super::{Link, OwningRef};
+use core::marker::PhantomData;
+use core::mem;
+use core::ops::DerefMut;
+#[cfg(test)]
+mod tests;
+
+//-----------------------------------------------------------------------------
+// Public API types
+//-----------------------------------------------------------------------------
+//  List
+/// An intrusive singly-linked list.
+///
+/// This type is a wrapper around a series of [`Node`]s. It stores [`Link`]s
+/// to the head and tail [`Node`]s and the length of the list.
+///
+/// # Type parameters
+/// - `T`: the type of the items stored by each `N`
+/// - `N`: the type of nodes in the list
+/// - `R`: the type of [`OwningRef`] that owns each `N`.
+///
+/// [`Node`]: trait.Node.html
+/// [`Link`]: ../struct.Link.html
+/// [`OwningRef]: ../trait.OwningRef.html
+#[derive(Default)]
+pub struct List<T, N, R> {
+    /// Link to the head node of the list.
+    head: Link<N>,
+
+    /// Length of the list.
+    len: usize,
+
+    /// Type marker for items stored in the list.
+    _elem_ty: PhantomData<T>,
+
+    /// Type marker for the `OwningRef` type.
+    _ref_ty: PhantomData<R>,
+}
+
+//  Linked
+/// Trait that must be implemented in order to be a member of an intrusive
+/// singly linked list.
+pub trait Linked: Sized {
+    /// Borrow this element's next [`Link`].
+    ///
+    /// [`Links`]: struct.Link.html
+    fn next(&self) -> &Link<Self>;
+
+    /// Mutably borrow this element's next [`Link`].
+    ///
+    /// [`Links`]: struct.Link.html
+    fn next_mut(&mut self) -> &mut Link<Self>;
+
+    /// De-link this node, returning its' next Link.
+    fn take_next(&mut self) -> Link<Self> {
+        mem::replace(self.next_mut(), Link::none())
+    }
+}
+
+//-----------------------------------------------------------------------------
+// Implementations
+//-----------------------------------------------------------------------------
+
+// ===== impl List =====
+
+impl<T, Node, R> List<T, Node, R> {
+    /// Create a new `List` with 0 elements.
+    pub const fn new() -> Self {
+        List {
+            head: Link::none(),
+            len: 0,
+            _elem_ty: PhantomData,
+            _ref_ty: PhantomData,
+        }
+    }
+
+    /// Returns the length of the list.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns true if the list is empty, false otherwise.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Borrows the first node of the list as an `Option`.
+    ///
+    /// Note that this is distinct from `front`: this method
+    /// borrows the head _node_, not the head _element_.
+    ///
+    /// # Returns
+    ///   - `Some(&N)` if the list has elements
+    ///   - `None` if the list is empty.
+    #[inline]
+    pub fn head(&self) -> Option<&Node> {
```

Comment:
Since the singly-linked list API surface is basically a stack, I'd consider renaming this to `peek` (and `head_mut` to `peek_mut`).